### PR TITLE
[AIRFLOW-2840] - add update connections cli option

### PR DESCRIPTION
**Pull request by caddac**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2840

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Adds an Update Connection CLI option. This is useful for updating connections in a CI/CD environment where deleting and adding connections could cause DAGS to fail.

### Tests

- [x] My PR adds the following unit tests:
My PR adds a `update connection` step to an existing unit test in `test/core.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by ashb** [comment]
connection ID is (by design) non-unique - its a mechanism of load balancing across multiple DBs for instance - you can have three conns with an ID of "mysql" and Airflow will pick a random one  each time.

Can you explicitly handle/detect this case (probably by refusing to update any of them?) and add tests that cover it too? Given it's relatively infrequent to do this I think it is okay to just refuse to update any of the conns in this case.

**Review by ashb — CHANGES_REQUESTED** [review]
connection ID is (by design) non-unique - its a mechanism of load balancing across multiple DBs for instance - you can have three conns with an ID of "mysql" and Airflow will pick a random one each time.

Can you explicitly handle/detect this case (probably by refusing to update any of them?) and add tests that cover it too? Given it's relatively infrequent to do this I think it is okay to just refuse to update any of the conns in this case.

**Comment by caddac** [comment]
Would updating all of the connections be acceptable? Maybe with a flag to indicate this behavior? Or is refusing to update preferable?

**Comment by caddac** [comment]
I've also noticed the cli option for adding connections doesn't support adding connections with duplicate `conn_id`, should it?

**Comment by bolkedebruin** [comment]
Please move the core of the code out of the cli (I.e. split between api/Commons and changes to the connection model) and make sure it works through the REST API as well. Direct access to the database should be limited in the cli as we will move away from this in the future.

**Comment by bolkedebruin** [comment]
O sorry @ashb I think I accidentally removed the label

**Comment by ashb** [comment]
@bolkedeburin no worries on the label, I was just using them as a way of tracking in the PRs page which ones we've looked at - entirely manual for now.

I think the default should refuse to update multiple, with an --update-all flag

**Comment by caddac** [comment]
@ashb @bolkedebruin  I've been working on refactoring the connection api and wanted to run the implementation by you before going too far. My thought is in `api/common/experimental/` there will be 4 files: `add_connection.py`, `list_connection.py`, `delete_connection.py`, `update_connection.py` where the DB access will occur. These functions will take explicitly defined arguments based on what they are doing (not dicts or something) and return `Connections` (or in the case of delete, `conn_id`(s)). 

Then each client (cli, local, json) would be responsible for parsing arguments and constructing meaningful print/return statements to the user (e.g. the cli client would formulate the message `Successfully added `conn_id`={conn_id} : {uri}` where as the json client would just return http 200 and the connection as json). 

thoughts?

**Comment by ashb** [comment]
Something like that sounds sensible - removing duplication is good. Don't forget about the Flask web UI(s - right now we still have two)

**Comment by bolkedebruin** [comment]
I wouldn't put the CRUD operations in separate files. Just use api/common/experimental/connections.py for simplicity

**Comment by caddac** [comment]
Can you give some feedback on the json_api? I'm not really sure what I should be doing there or how to test it. Thanks!

File: airflow/api/client/local_client.py [hunk 1]
```diff
@@ -51,3 +51,83 @@ def create_pool(self, name, slots, description):
...
+        :param conn_type:
+        :param conn_host:
+        :param conn_login:
+        :param conn_password:
+        :param conn_schema:
+        :param conn_port:
+        :param conn_extra:
+        :return: The new Connection
+        """
+        return connections.add_connection(conn_id,
```
Comment (ashb) [inline airflow/api/client/local_client.py:77]: We probably shouldn't return objects over the API  (as only the "Local" api can do this - we couldn't return an object over JSON for instance) - this should be the conn_id that we return I think.

It is possibly worth returning the numeric ID (the `id` column, not to be confused with the `conn_id` column!) as this would let us delete individual connections if there are multiple ones with the same conn_id.

Similarly for the other APIs- ids or JSON representations of the conn.

**Review by ashb — COMMENTED** [review]
Haven't had time to give this a full re-review yet, sorry.

**Comment by ron819** [comment]
Regarding @ashb  first comment from Aug about multiple connections. There is also a bug to consider: https://issues.apache.org/jira/browse/AIRFLOW-2784    Your tests can't assume that all conn_id with same name are for the same conn type until this bug is fixed.

File: airflow/api/client/local_client.py (hunk 1 above)
Comment (caddac) [reply airflow/api/client/local_client.py:77]: So the idea here was the client(s) would return the same objects (regardless of client type) and leave formatting the response to the api (cli or http). What about returning the json representation from the client, (which the local and json clients could do) and still leaving formatting to the API. TBH, the differences in clients confuses me and I'm still not sure what  or why we have two different clients given that the end users seem to interact with the API rather than the client directly. Any clarity you could provide there would be appreciated.

**Comment by caddac** [comment]
@ron819 I'm going to address this bug in my PR. 

@ashb I'm still looking for some clarity on what is going on with the json_client. Even looking at the `pool` examples that currently exist there, I'm still not sure what exactly I need to do. Seems like some kind of http handling, but I assumed that was being handled by the `www` and `www_rbac` code. Any guidance you can give would be great.

**Comment by caddac** [comment]
@ashb Think I got the json_client figured out, will try to get my changes committed tonight.

**Comment by ashb** [comment]
Great! Sorry I didn’t respond - paid work has been busy :)

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -1110,133 +1108,77 @@ def version(args):
...
-            msg = tabulate(conns, ['Conn Id', 'Conn Type', 'Host', 'Port',
-                                   'Is Encrypted', 'Is Extra Encrypted', 'Extra'],
-                           tablefmt="fancy_grid")
-            if sys.version_info[0] < 3:
-                msg = msg.encode('utf-8')
-            print(msg)
-            return
+        conns = api_client.list_connections()
+        # format it for the cli
+        conns = list(map(lambda c: (c['conn_id'], c['conn_type'],
```
Comment (ashb) [inline airflow/bin/cli.py]: Minor point: we don't need the `list()` here

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -1110,133 +1108,77 @@ def version(args):
...
-                deleted_conn_id = to_delete.conn_id
-                session.delete(to_delete)
-                msg = '\n\tSuccessfully deleted `conn_id`={conn_id}\n'
-                msg = msg.format(conn_id=deleted_conn_id)
-                print(msg)
-            return
+        try:
+            print(api_client.delete_connection(args.conn_id, args.delete_all))
+        except MissingArgument as ma:
+            print(ma)
```
Comment (ashb) [inline airflow/bin/cli.py:1131]: We should exit with a Non-zero code here.

File: airflow/bin/cli.py [hunk 4]
```diff
@@ -1110,133 +1108,77 @@ def version(args):
...
-        if missing_args:
-            msg = ('\n\tThe following args are required to add a connection:' +
-                   ' {missing!r}\n'.format(missing=missing_args))
-            print(msg)
-        if invalid_args:
-            msg = ('\n\tThe following args are not compatible with the ' +
-                   '--add flag and --conn_uri flag: {invalid!r}\n')
-            msg = msg.format(invalid=invalid_args)
+        try:
+            new_conn = api_client.add_connection(
```
Comment (ashb) [inline airflow/bin/cli.py:1136]: Might be easier to do `new_conn = api_client.add_connection(vars(args))`? (that does rely on the names from argparser matching the named args in the api client, but they do right now)

Not sure on this one.

File: airflow/bin/cli.py [hunk 5]
```diff
@@ -1110,133 +1108,77 @@ def version(args):
...
+                conn_type=args.conn_type,
+                conn_host=args.conn_host,
+                conn_login=args.conn_login,
+                conn_password=args.conn_password,
+                conn_schema=args.conn_schema,
+                conn_port=args.conn_port,
+                conn_extra=args.conn_extra)
+
+        except MissingArgument as ma:
+            print(ma)
```
Comment (ashb) [inline airflow/bin/cli.py:1148]: Non-zero exit code here too

File: tests/api/common/experimental/connection_tests.py [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+                                       conn_type="not_postgres",
+                                       conn_host="host",
+                                       conn_login="airflow",
+                                       conn_password="airflow",
+                                       conn_schema="airflow",
+                                       conn_port="5432")
+            self.assertEqual(ia.exception, "Connections with the same id must all be of the same type")
+
+        # validate expected connections reached the DB
+        session = settings.Session()
```
Comment (ashb) [inline tests/api/common/experimental/connection_tests.py]: Use `self.session`?

File: tests/api/common/experimental/connection_tests.py [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+                      .query(models.Connection)
+                      .filter(models.Connection.conn_id == conn_id)
+                      .first())
+            result = (result.conn_id, result.conn_type, result.host,
+                      result.port, result.get_extra())
+            if conn_id in ['new1', 'new2', 'new3']:
+                self.assertEqual(result, (conn_id, 'postgres', 'host', 5432,
+                                          extra[conn_id]))
+            elif conn_id == 'new4':
+                print('result: ', result)
```
Comment (ashb) [inline tests/api/common/experimental/connection_tests.py]: Generally having bare prints in tests should be avoided - it gets noisy over time

File: tests/api/common/experimental/connection_tests.py [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+            conn_port="5432"
+        ), Connection(
+            conn_id="new2",
+            conn_type="postgres",
+            host="host",
+            login="airflow",
+            password="airflow",
+            schema="airflow",
+            port="5432"
+        ))
```
Comment (ashb) [inline tests/api/common/experimental/connection_tests.py:244]: From a quick look it appears that all the test cases use this connection - could we create it once in setUp() to reduce the amount of code in each test function?

File: tests/core.py [hunk 9]
```diff
@@ -1188,33 +1218,13 @@ def test_cli_connections_list(self):
...
-                ['connections', '--list', '--conn_id=fake', '--conn_uri=fake-uri',
-                 '--conn_type=fake-type', '--conn_host=fake_host',
-                 '--conn_login=fake_login', '--conn_password=fake_password',
-                 '--conn_schema=fake_schema', '--conn_port=fake_port', '--conn_extra=fake_extra']))
-            stdout = mock_stdout.getvalue()
-
-        # Check list attempt stdout
-        lines = [l for l in stdout.split('\n') if len(l) > 0]
-        self.assertListEqual(lines, [
-            ("\tThe following args are not compatible with the " +
```
Comment (ashb) [inline tests/core.py:1213]: What has happened to this test case? I think, it would still be useful, but I don't see it anymore?

**Review by ashb — CHANGES_REQUESTED** [review]
A few small changes/comments. Looking good overall.

One thing that wasn't clear from a read of the code: does the API return the passwords on list or not? (I don't think it should, as right now none of the current methods of viewing connection in the CLI or WebUI give the value of the password)

File: tests/api/common/experimental/connection_tests.py [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+        super(ConnectionTests, self).tearDown()
+
+    @staticmethod
+    def _cleanup(session=None):
+        if session is None:
+            session = Session()
+
+        session.query(models.Pool).delete()
+        session.query(models.Variable).delete()
+        session.query(models.Connection).delete()
```
Comment (ashb) [inline tests/api/common/experimental/connection_tests.py]: This is causing your test failures:

>    AirflowException: The conn_id `sqlite_default` isn't defined

from elsewhere in the tests.

File: airflow/bin/cli.py (hunk 3 above)
Comment (caddac) [reply airflow/bin/cli.py:1131]: is raising the exception good enough? or is there a specific way to exit with non-zero code?

Comment (ashb) [reply airflow/bin/cli.py:1131]: Raising the `ma` exception would work but would provide a messy output (i.e. it would include the stack trace which is not useful here):

https://github.com/apache/airflow/blob/9f3eae5e82aef4550027e5c3163715dad67bdb57/airflow/bin/cli.py#L1295-L1296

```
        except MissingArgument as ma:
            raise SystemExit(ma)
```

Should probably do the trick (and not include any errant stack-traces. If that still includes a stack trace, or errors then `SystemExit(str(ma)))`

File: tests/core.py (hunk 9 above)
Comment (caddac) [reply tests/core.py:1213]: This test case was removed because the behavior of rejecting arguments for the list command was removed. It just lists the connections no matter what args are provided now.

**Comment by caddac** [comment]
I see 5 tests failing, 1 of which is my `test_delete_connection` test. This test succeeds locally (on docker) for both python 2.7 and 3.6. The other tests also fail locally for me (on docker) from master. Not sure where to go from here, any suggestions @ashb ?

File: tests/core.py [hunk 11]
```diff
@@ -1248,75 +1258,166 @@ def test_cli_connections_add_delete(self):
...
-            ("\tSuccessfully added `conn_id`=new5 : " +
-             "hive_metastore://airflow:airflow@host:9083/airflow"),
-            ("\tSuccessfully added `conn_id`=new6 : " +
-             "google_cloud_platform://:@:")
+            "Successfully added `conn_id`=new1 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new2 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new3 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new4 : postgres://airflow:airflow@host:5432/airflow",
+            "Successfully added `conn_id`=new5 : hive_metastore://airflow:airflow@host:9083/airflow",
+            "Successfully added `conn_id`=new6 : google_cloud_platform://:@:",
```
Comment (mik-laj) [inline tests/core.py:1278]: According to the "RFC 3986: Uniform Resource Identifier (URI): Generic Syntax" specification, this is not a valid address.
Schema part must be compatible with the format:
````
      scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
````
I created a issue https://issues.apache.org/jira/browse/AIRFLOW-3616  and working on PR: https://github.com/PolideaInternal/airflow/pull/30 
which deals with this topic.

**Review by mik-laj — CHANGES_REQUESTED** [review]
Please improve compatibility with RFC 3986 Section 3.1

File: airflow/api/client/json_client.py [hunk 12]
```diff
@@ -89,3 +89,84 @@ def delete_pool(self, name):
...
+        :param conn_host:
+        :param conn_login:
+        :param conn_password:
+        :param conn_schema:
+        :param conn_port:
+        :param conn_extra:
+        :return: The updated Connection
+        """
+        endpoint = '/api/experimental/connections/'
+        url = urljoin(self._api_base_url, endpoint)
```
Comment (mik-laj) [inline airflow/api/client/json_client.py:161]: This should be done in the `_request` method. Can you correct this or will we do it in a separate PR?

File: airflow/www/api/experimental/endpoints.py [hunk 13]
```diff
@@ -253,3 +254,70 @@ def delete_pool(name):
...
+def update_connection():
+    """Delete pool."""
+    params = request.get_json(force=True)
+    try:
+        new_conn = connection_api.update_connection(**params)
+    except AirflowException as err:
+        _log.error(err)
+        response = jsonify(error="{}".format(err))
+        response.status_code = err.status_code
+        return response
```
Comment (mik-laj) [inline airflow/www/api/experimental/endpoints.py:321]: This code snippet is repeated in any method. Should we think about how to avoid repeating this piece of code? Maybe introducing a decorator?

**Comment by caddac** [comment]
will reopen when working
